src/test/compile-fail/move-trait-leaks-in-impl-trait.rs

```diff
@@ -0,0 +1,33 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(conservative_impl_trait, immovable_types)]
+
+use std::marker::{Move, Immovable};
+
+trait Foo: ?Move {}
+
+impl<T: ?Move> Foo for T {}
+
+fn immovable() -> impl Foo {
+    Immovable
+}
+
+fn movable() -> impl Foo {
+    true
+}
+
+fn test<T>(_: T) {}
+
+fn main() {
+    test(immovable());
+    //~^ ERROR the trait bound `std::marker::Immovable: std::marker::Move` is not satisfied
```

Comment:
hmm. This is certainly fine as a conservative choice for now (I do see that the RFC says "Existential types (impl Trait) are Move if their underlying type are Move" which I take to be an "if *and only if*").

But I find it interesting because I would have guessed that since no borrows are taken of the value returned from `immovable()`, nor do we call any methods on it that take `&self`/`&mut self`, that it would be safe to subsequently move it in the call to `test(..)`

Reply:
It is safe to move it into `test`. The problem is that the `test` function has a `Move` bound, so it may rely on the type being movable after taking the address of it.